src/repository.c

```diff
@@ -846,32 +938,66 @@ int git_repository_open_ext(
 	if (config && (error = check_repositoryformatversion(config)) < 0)
 		goto cleanup;
 
-	if ((flags & GIT_REPOSITORY_OPEN_BARE) != 0)
+	if ((opts.flags & GIT_REPOSITORY_OPEN_BARE) != 0)
 		repo->is_bare = 1;
 	else {
 
 		if (config &&
 		    ((error = load_config_data(repo, config)) < 0 ||
-		     (error = load_workdir(repo, config, &workdir)) < 0))
+		     (error = load_workdir(repo, config, &env.workdir_path)) < 0))
```

Comment:
We move on to actually setup the working directory, at https://github.com/libgit2/libgit2/pull/4856/files#diff-7cbc97c802e21b199f176b493cec8e9cR300, and we pass the workdir identified by the "find" step. This is the part where the crash I'm getting comes from, and I'm starting to get completely lost (and I can't inline-review that code anyway, so there goes) :

- `core.worktree` is loaded from the config, as it can now be made use of.
- if `repo->is_worktree` is true (it only performed filesystem-checks), we read `repo->gitdir` + "gitdir" (already loaded while finding in `env->gitlink`), `dirname()` that, and use the result as our workdir.
- if `core.worktree` is set, we check its value and use that as our workdir.
- if the `parent_path` is passed, we check that it's a directory and use that.
- finally, we `dirname(repo->gitdir)` and use that.

It seems that `core.worktree` gets a say iif the repository structure is slightly "wrong" — a `$GIT_DIR/gitdir` file that doesn't point to a valid repo. The fact that it is passing as a parameter the very thing this code should be setting up doesn't help 😉.